# Atividade Prática — QA com APIs (SWAPI)

## Versão do aluno — com scripts base

**Tempo estimado:** 60 minutos  
**Ferramenta:** Google Colab  
**Formato:** em grupos de **2 a 4 componentes** ou, se preferirem, **individualmente**.  
**Objetivo:** investigar respostas de uma API, ajustar scripts simples e responder às perguntas como um QA.

### Instruções
Nesta atividade, vocês **não precisam criar tudo do zero**.  
Os scripts já estão prontos como base. A tarefa do grupo será:

1. **Executar** os blocos de código
2. **Alterar apenas o que for pedido**
3. **Observar os resultados**
4. **Responder às perguntas em Markdown**
5. **Discutir em grupo** quais validações um QA faria

### Entrega
Cada grupo pode entregar **um único notebook** com:
- nomes dos integrantes
- respostas das perguntas
- scripts ajustados
- asserts completados

Se alguém optar por fazer sozinho, pode entregar individualmente sem problema.


In [1]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/10/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/10/"
    },
    "https://swapi.dev/api/people/13/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/13/"
    },
    "https://swapi.dev/api/planets/2/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/2/"
    },
    "https://swapi.dev/api/starships/10/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/10/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


## Identificação do grupo

- Integrante 1: Luis Gustavo dos Santos Talgatti

## Parte 1 — Explorando a API

Nesta primeira parte, vocês vão consultar 4 endpoints.  
O trabalho do grupo será **alterar apenas a URL** em cada bloco, executar e observar o retorno.


### 1.1 Personagem 47

**Tarefa:** execute o script e observe:
- status code
- nome
- altura
- cor dos olhos

In [2]:
url = "https://swapi.dev/api/people/47/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Ratts Tyerel",
    "height": "79",
    "mass": "15",
    "hair_color": "none",
    "skin_color": "grey, blue",
    "eye_color": "unknown",
    "birth_year": "unknown",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/38/",
    "films": [
        "https://swapi.dev/api/films/4/"
    ],
    "species": [
        "https://swapi.dev/api/species/16/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-20T09:53:15.086000Z",
    "edited": "2014-12-20T21:17:50.410000Z",
    "url": "https://swapi.dev/api/people/47/"
}


**Respostas do aluno — Personagem 47**

- Status code: 200
- Nome do personagem: Ratts Tyerel
- Altura: 79 cm
- Cor dos olhos: Desconhecido
- O JSON parece completo? Sim

### 1.2 Personagem 77

**Tarefa:** neste bloco, troque apenas o endpoint se necessário, execute e observe o retorno.

In [3]:
url = "https://swapi.dev/api/people/77/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "San Hill",
    "height": "191",
    "mass": "unknown",
    "hair_color": "none",
    "skin_color": "grey",
    "eye_color": "gold",
    "birth_year": "unknown",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/57/",
    "films": [
        "https://swapi.dev/api/films/5/"
    ],
    "species": [
        "https://swapi.dev/api/species/34/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-20T17:58:17.049000Z",
    "edited": "2014-12-20T21:17:50.484000Z",
    "url": "https://swapi.dev/api/people/77/"
}


**Respostas do aluno — Personagem 77**

- Status code: 200
- Nome do personagem: San Hill
- Altura: 191 cm
- Cor dos olhos: Dourado
- O JSON parece completo? Sim

### 1.3 Planeta 1

**Tarefa:** execute o bloco e observe:
- nome do planeta
- clima
- terreno
- população

In [4]:
url = "https://swapi.dev/api/planets/1/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Tatooine",
    "rotation_period": "23",
    "orbital_period": "304",
    "diameter": "10465",
    "climate": "arid",
    "gravity": "1 standard",
    "terrain": "desert",
    "surface_water": "1",
    "population": "200000",
    "residents": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/6/",
        "https://swapi.dev/api/people/7/",
        "https://swapi.dev/api/people/8/",
        "https://swapi.dev/api/people/9/",
        "https://swapi.dev/api/people/11/",
        "https://swapi.dev/api/people/43/",
        "https://swapi.dev/api/people/62/"
    ],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/"
    ],
    "created": "2014-12-09T13

**Respostas do aluno — Planeta 1**

- Status code: 200
- Nome do planeta: Tatooine
- Clima: Arido
- Terreno: Deserto
- População: 200000
- O JSON parece completo? Sim

### 1.4 Nave 3

**Tarefa:** execute o bloco e observe:
- nome da nave
- modelo
- fabricante
- velocidade máxima em atmosfera

In [7]:
url = "https://swapi.dev/api/starships/3/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Star Destroyer",
    "model": "Imperial I-class Star Destroyer",
    "manufacturer": "Kuat Drive Yards",
    "cost_in_credits": "150000000",
    "length": "1,600",
    "max_atmosphering_speed": "975",
    "crew": "47,060",
    "passengers": "n/a",
    "cargo_capacity": "36000000",
    "consumables": "2 years",
    "hyperdrive_rating": "2.0",
    "MGLT": "60",
    "starship_class": "Star Destroyer",
    "pilots": [],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/"
    ],
    "created": "2014-12-10T15:08:19.848000Z",
    "edited": "2014-12-20T21:23:49.870000Z",
    "url": "https://swapi.dev/api/starships/3/"
}


**Respostas do aluno — Nave 3**

- Status code: 200
- Nome da nave: Star Destroyer
- Modelo: Imperial I-class Star Destroyer
- Fabricante: Kuat Drive Yards
- Velocidade máxima em atmosfera: 975
- O JSON parece completo? Sim

## Parte 2 — Extraindo campos específicos

Agora você vai executar os scripts e **completar as respostas**.

### 2.1 Extraindo campos do personagem

**Tarefa:** execute o bloco e depois responda.

In [11]:
url = "https://swapi.dev/api/people/4/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Darth Vader
Altura: 202
Cor dos olhos: yellow
Gênero: male


**Perguntas**
1. Qual campo foi usado para mostrar o nome? "name"
2. Qual campo foi usado para mostrar a altura? "height"
3. Se você fosse QA, quais desses campos validaria primeiro? O campo "name"

### 2.2 Extraindo campos do planeta

**Tarefa:** execute o bloco e complete as respostas.

In [12]:
url = "https://swapi.dev/api/planets/1/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Tatooine
Clima: arid
Terreno: desert
População: 200000


**Perguntas**
1. O campo população parece numérico? "population"
2. O clima foi retornado corretamente? Sim
3. Se você fosse QA, o que validaria nesse planeta? O campo "name"

## Parte 3 — Pensando como QA

Agora você vai trabalhar com validações simples.

### 3.1 Validações básicas

**Tarefa:** execute o bloco abaixo e depois escreva o que cada validação verifica.

In [14]:
url = "https://swapi.dev/api/people/7/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


**Explique com suas palavras**
1. O que a validação 1 verifica? Verifica se a requisição foi bem-sucedida e a API respondeu corretamente (HTTP 200 OK)
2. O que a validação 2 verifica? O campo "name"
3. O que a validação 3 verifica? Se o campo "name" está vazio
4. O que a validação 4 verifica? Se o campo "height" existe

### 3.2 Complete as validações

**Tarefa:** execute o bloco abaixo. Ele já está funcionando. Depois, alterem as expressões se o professor pedir outra nave ou outro endpoint. Em seguida, expliquem por que cada validação faz sentido para QA.


In [15]:
url = "https://swapi.dev/api/starships/3/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200) # Garante a disponibilidade do serviço. Se não retornar 200, a API pode estar fora do ar ou com erro interno.

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos.
print("Campo 'name' existe:", "name" in dados) # Garante o contrato/schema da API. Verifica se os desenvolvedores não removeram ou renomearam um campo obrigatório acidentalmente.
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "") # Garante a qualidade do dado. Não basta o campo existir, ele precisa conter uma informação válida e não vir em branco.


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


## Parte 4 — Testes automatizados com assert

Agora vocês vão trabalhar sobre scripts já prontos. Os asserts abaixo já executam sem erro. A tarefa do grupo é **entender o que cada linha valida**, ajustar se o professor pedir outro recurso e registrar as respostas.


### 4.1 Assert no personagem

**Tarefa:** execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.


In [17]:
url = "https://swapi.dev/api/people/13/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado" # Valida se a chamada teve sucesso.
assert "name" in dados, "Erro: campo name não encontrado" # Valida a estrutura do JSON, checando se a chave "name" está presente.
assert dados["name"] != "", "Erro: nome vazio" # Valida o preenchimento do dado, garantindo que o nome não é uma string vazia.

print("Todos os testes do personagem passaram!")


Todos os testes do personagem passaram!


### 4.2 Assert na nave

**Tarefa:** execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.


In [18]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

print("Todos os testes da nave passaram!")


Todos os testes da nave passaram!


**Qual campo foi validado em cada assert?**

Foi validado o status da requisição, a existência do campo manufacturer (fabricante) e se o campo model (modelo) não está vazio.

**Proposta de validação extra:**

assert "passengers" in dados, "Erro: campo passageiros não encontrado" (Valida se a chave de passageiros existe na resposta).

## Parte 5 — Cenário negativo

Agora você vai observar como a API se comporta quando pedimos um recurso inexistente.

In [20]:
url = "https://swapi.dev/api/people/0000/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


**Perguntas**
1. Qual status code foi retornado? 404
2. Esse comportamento está correto? Sim
3. Por que um QA deve testar cenários negativos? Para garantir a resiliência do sistema. É preciso validar se a API lida bem com erros de usuários, se retorna o status code adequado e se não expõe dados sensíveis do código no erro.

## Parte 6 — Mini desafio final

Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [22]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/species/3/")

Origem dos dados: online
Status Code: 200
name: Wookie
classification: mammal
designation: sentient
average_height: 210
skin_colors: gray
hair_colors: black, brown
eye_colors: blue, green, yellow, brown, golden, red
average_lifespan: 400
homeworld: https://swapi.dev/api/planets/14/
language: Shyriiwook
people: ['https://swapi.dev/api/people/13/', 'https://swapi.dev/api/people/80/']
films: ['https://swapi.dev/api/films/1/', 'https://swapi.dev/api/films/2/', 'https://swapi.dev/api/films/3/', 'https://swapi.dev/api/films/6/']
created: 2014-12-10T16:44:31.486000Z
edited: 2014-12-20T21:36:42.142000Z
url: https://swapi.dev/api/species/3/


**Respostas finais**
1. Qual endpoint você testou no desafio? https://swapi.dev/api/species/3/
2. Qual foi o status code? 200
3. Qual campo chamou mais a sua atenção? "average_lifespan"
4. Se você fosse QA, quais 3 validações faria nesse recurso? Validaria se o status code é 200, validaria se o campo name existe e é igual a "Wookiee" e validaria se o tipo de dado de language é uma string e não está vazio.

# Entrega da atividade

Entreguem:

1. O notebook preenchido  
2. As respostas em Markdown  
3. Os scripts ajustados  
4. Os asserts analisados e/ou adaptados  
5. Uma conclusão curta do grupo: **quais foram as 3 validações mais importantes?**

## Fechamento
Ao final desta atividade, vocês praticaram:

- leitura de resposta JSON
- análise de status code
- validação de campos
- pensamento de QA
- testes simples com assert
- análise de cenário negativo


Quais foram as 3 validações mais importantes?

Status Code (200): A base de tudo, confirma que a comunicação com o servidor funcionou.

Validação de Contrato (chaves existem): Garante que a estrutura do JSON devolvido é a que o Front-end espera receber.

Validação de Dados (campos não vazios): Garante que o banco de dados está enviando informações úteis e não apenas campos em branco.